### PR TITLE
ENG-4273: uninstalls orphaned components during installation

### PR DESCRIPTION
**Pull request by antromeo**

### CODE REVIEW

File: src/main/java/org/entando/kubernetes/service/digitalexchange/job/EntandoBundleInstallService.java [hunk 1]
```diff
@@ -332,6 +334,13 @@ private String uninstallOrphanedComponents(EntandoBundleJobEntity parentJob, Ent
                 Queue<Installable> bundleInstallableComponentsDiff = getBundleInstallableComponents(latestBundleReader, conflictStrategy, diff);
 
                 Queue<EntandoBundleComponentJobEntity> componentJobQueueDiff = bundleInstallableComponentsDiff.stream()
+                        .filter(i -> {
```
Comment (antromeo) [inline src/main/java/org/entando/kubernetes/service/digitalexchange/job/EntandoBundleInstallService.java]: @firegloves note that resources in file browser are preserved

File: src/main/java/org/entando/kubernetes/service/digitalexchange/job/EntandoBundleInstallService.java [hunk 2]
```diff
@@ -311,6 +324,83 @@ private CompletableFuture<Void> submitInstallAsync(EntandoBundleJobEntity parent
...
+        List<EntandoBundleComponentJobEntity> componentJobListDiff = new ArrayList<>(componentJobQueueDiff);
+        return usageService.getComponentsUsageDetails(componentJobListDiff).stream()
+                .filter(componentUsage -> {
+                    if (!componentUsage.isExist() || componentUsage.getHasExternal()) {
+                        log.debug(String.format(
+                                "Component '%s' is not found or contains external references so it can't be uninstalled",
+                                componentUsage.getCode()));
+                        return false;
+                    }
+                    return true;
```
Comment (firegloves) [inline src/main/java/org/entando/kubernetes/service/digitalexchange/job/EntandoBundleInstallService.java]: Please consider extracting this function to a method with a comprehensive name just to ease the reading process

File: src/main/java/org/entando/kubernetes/service/digitalexchange/job/EntandoBundleInstallService.java [hunk 3]
```diff
@@ -311,6 +324,83 @@ private CompletableFuture<Void> submitInstallAsync(EntandoBundleJobEntity parent
...
+
+    private static Queue<EntandoBundleComponentJobEntity> getBundleComponentJobEntityQueue(EntandoBundleJobEntity parentJob,
+                                                                                           Queue<Installable> bundleInstallableComponentsDiff) {
+        return bundleInstallableComponentsDiff.stream()
+                .filter(i -> {
+                    if (i.getRepresentation() instanceof WidgetDescriptor) {
+                        WidgetDescriptor wd = (WidgetDescriptor) i.getRepresentation();
+                        return TYPE_WIDGET_APPBUILDER.equals(wd.getType());
+                    }
+                    return false;
```
Comment (firegloves) [inline src/main/java/org/entando/kubernetes/service/digitalexchange/job/EntandoBundleInstallService.java]: Please consider extracting this function to a method with a comprehensive name just to ease the reading process. Might it also be worth adding a sort of `isTypeAppBuilder` within the WidgetDescriptor itself?

File: src/main/java/org/entando/kubernetes/service/digitalexchange/job/EntandoBundleInstallService.java [hunk 4]
```diff
@@ -311,6 +324,83 @@ private CompletableFuture<Void> submitInstallAsync(EntandoBundleJobEntity parent
...
+                })
+                .map(i -> {
+                    EntandoBundleComponentJobEntity cj = new EntandoBundleComponentJobEntity();
+                    cj.setParentJob(parentJob);
+                    cj.setComponentType(i.getComponentType());
+                    cj.setComponentId(i.getName());
+                    cj.setChecksum(i.getChecksum());
+                    cj.setInstallable(i);
+                    cj.setAction(i.getAction());
+                    return cj;
```
Comment (firegloves) [inline src/main/java/org/entando/kubernetes/service/digitalexchange/job/EntandoBundleInstallService.java]: This code should be already present in the same file, please centralize it to a designated method.

File: src/main/java/org/entando/kubernetes/service/digitalexchange/job/EntandoBundleInstallService.java [hunk 5]
```diff
@@ -545,16 +741,50 @@ private boolean isUninstallable(EntandoBundleComponentJobEntity component) {
                 && component.getComponentType() == ComponentType.PLUGIN);
     }
 
-    private JobResult executeRollback(Installable<?> installable) {
+    // TODO: moves the inner class
```
Comment (firegloves) [inline src/main/java/org/entando/kubernetes/service/digitalexchange/job/EntandoBundleInstallService.java]: can we delete this TODO?

**Review by firegloves — CHANGES_REQUESTED** [review]

**Review by firegloves — APPROVED** [review]
LGTM

**Pull request merged**